### ISSUE TITLE
Can anyone give some advice related to image classification using tiny cnn??

### ISSUE THREAD

**Issue by higrepyo** [tiny-dnn/tiny-dnn#211]
I'm currently trying to construct a convnet which can classify vehicle-rear images vs. everything else. 

I've gone through the MNIST and cifar-10 example provided, but the results of the trained network isn't good.

 I used the exact same network of the cifar-10 example(tried on the MNIST net also), using about 2k positive vehicle rear images and 4k negative images which were cut and saved randomly from highway scenes(such as lanes, fences, trees, etc.).

I used mnisten to make my dataset to 32x32 grayscale images, and 1000 testing datasets.

The weird thing is that I get almost over 98% on the test results, but when I actually test them using the test.cpp file of the MNIST example(checked it works with the number "4" provided), the trained network doesn't seem to work.

I've worked almost 2 weeks on this and this phenomenon is driving me crazy. 

I tried training with smaller datasets, larger datasets, and also tried to divide the negative datasets into several categories such as right lane, left lane, fences. Every time, it gives a promising test result(at least over 95%), but it always doesn't work on actual tests.

What was worse is that even when I performed tests on the training images, it didn't give the right output. 

Does anyone have any suggestions? I'll be watching this thread 24/7.

**Comment by dicroce** [comment]
I strongly suspect that some of your test data is slipping into your training data. I made the same mistake and was getting accuracy's in the high 90's. I believe the highest I got was in the 70's (against cifar10) after I fixed that.

What I find frustrating is that convnet.js beats tiny-cnn within a fairly small number of epochs!

**Comment by higrepyo** [comment]
@dicroce 

I also get 70~80% for the cifar-10 example. That was the reason I believed tiny cnn worked properly.

The problem I'm talking about happens when I make my own dataset using mnisten(mnisten -d images -o my_dataset -s 32x32 -n 1000 in terminal).

Would this cause the problem you're mentioning?

**Comment by nyanp** [comment]
@higrepyo 
Thank you for using tiny-cnn :)

Did you change ```construct_net``` function to fit the cifar-10 networks? We must use the same network model between train and test.

**Comment by higrepyo** [comment]
@nyanp 
Yes, I used network<cross_entropy, adam> for both training and testing.

P.S. Thank you for the wonderful c++ header only framework :)

**Comment by nyanp** [comment]
@higrepyo 
Ok. Could you attach your test.cpp, trained model file, and one of your test image?

**Comment by higrepyo** [comment]
@nyanp 
I've never uploaded a file before, so I'll send you a .zip file of the c++ code, 1 positive sample, and 1 negative sample.

 In the main code, MODE is set to 3 for testing(currently set to 3) and set to 2 for training. 

Thank you for your assistance
[myCNN.zip](https://github.com/nyanp/tiny-cnn/files/344620/myCNN.zip)

**Comment by nyanp** [comment]
@higrepyo 
Thanks, could you send me idx files for training/testing? I've checked your code but I can't find a problem...

**Comment by higrepyo** [comment]
@nyanp 

Here there are. Once again, thank you for looking in to this problem.
[idx_files.zip](https://github.com/nyanp/tiny-cnn/files/345147/idx_files.zip)

**Comment by nyanp** [comment]
@higrepyo 
There is a mismatching of value-range between training data and acual test data. Since the original MNIST data is encoded as "black on white" digits, this example negate the input image to handle "white on black" digit image (4.bmp).

Would you change the following line 

```cpp
	// mnist dataset is "white on black", so negate required
	std::transform(resized.begin(), resized.end(), std::back_inserter(data),
		[=](uint8_t c) { return (255 - c) * (maxv - minv) / 255.0 + minv;});

```

to

```cpp
	// mnist dataset is "white on black", so negate required
	std::transform(resized.begin(), resized.end(), std::back_inserter(data),
		[=](uint8_t c) { return c * (maxv - minv) / 255.0 + minv; });
```

and tell me the result? I hope this will help you :)

**Comment by higrepyo** [comment]
@nyanp 
I'm on a business trip so I'll do it first thing tomorrow after I arrive.

Thank you :)

**Comment by higrepyo** [comment]
@nyanp 
OMG it completely works !

Now knowing the problem, I'm very curious how it even slightly worked before I changed the code you mentioned :-o


P.S.

Thank you so much. Is there a way to cite tiny cnn in papers??

I really want to thank tiny cnn and you after I finish the current project I'm working on.

**Comment by nyanp** [comment]
@higrepyo
I'm glad to hear that ;) Although we don't have a paper for citations, so you can cite the github repo. For example:

@misc{nomi2016,
author = {Taiga Nomi },
title = {tiny-cnn},
year = {2016},
publisher = {GitHub},
journal = {GitHub repository},
howpublished = {\url{https://github.com/nyanp/tiny-cnn}},
commit = {aa71c81}
}

I'd very glad if you tell me your paper after you finish the project :)

**Comment by Randl** [comment]
@higrepyo Are you finished yet? Can you give us a link to the paper so we can mention it in our README? :)

**Comment by higrepyo** [comment]
I'm sorry to notice you that I dropped out of the project 6 months ago because I got a full time job.

--------- 원본 메일 ---------

    
        보낸사람: Evgeniy Zheltonozhskiy <notifications@github.com>
        받는사람: tiny-dnn/tiny-dnn <tiny-dnn@noreply.github.com>
        
            참조: higrepyo <higre_pyo@hanmail.net>, Mention <mention@noreply.github.com>
        
        날짜: 2017년 2월 8일 수요일, 08시 51분 14초 +0900
        제목: Re: [tiny-dnn/tiny-dnn] Can anyone give some advice related to image classification using tiny cnn?? (#211)
        
        
    
    
        @higrepyo Are you finished yet? Can you give us a link to the paper so we can mention it in our README? :)

**Comment by Randl** [comment]
@higrepyo that's a pity. I'm clothing the issue then.
